File: SortVision/public/code/insertion/golang/insertionSort.go
func InsertionSort(arr []int) []int {
    // Get the length of input array
    n := len(arr)
    
    // Iterate through array starting from the second element
    for i := 1; i < n; i++ {
        // Store current element as key to be inserted
        // in correct position
        key := arr[i]
        
        // Initialize j as the element before current position
        j := i - 1
        
        // Move elements of arr[0..i-1] that are greater than key
        // one position ahead of their current position
        // This creates the space to insert the key
        for j >= 0 && arr[j] > key {
            // Shift elements to the right
            arr[j+1] = arr[j]
            j = j - 1
        }
        
        // Place key in its correct position
        // j+1 is the correct position because j was decremented
        // one extra time in the last iteration of the inner loop
        arr[j+1] = key
    }
    
    // Return the sorted array
    return arr
}

File: SortVision/public/code/insertion/java/insertionSort.java

    public static void insertionSort(int[] arr) {
        // Iterate through the array starting from the second element
        for (int i = 1; i < arr.length; i++) {
            // Store the current element that needs to be inserted in the right position
            int key = arr[i];
            
            // Initialize j as the position before i
            int j = i - 1;
            
            // Move elements of arr[0..i-1] that are greater than key
            // to one position ahead of their current position
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];    // Shift element to the right
                j = j - 1;              // Move to previous position
            }
            
            // Place the key in its correct position
            arr[j + 1] = key;
        }
    }

   

